akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala

```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
+              continue(input drop emitEnd, 0)(parseEntity(null, null, simpleEmit, simpleEmit))
+            } else continue(input, offset)(parseEntity(headers, contentType, emitPartChunk, emitFinalPartChunk))
+        }
+
+      def emit(bytes: ByteString): Unit = if (bytes.nonEmpty) emit(EntityPart(bytes))
+
+      def emit(element: Output): Unit = output = output.enqueue(element)
+
+      def dequeue(): Output = {
+        val head = output.head
+        output = output.tail
+        head
       }
-      rec(offset)
-    } catch {
-      case NotEnoughDataException ⇒
-        // we cannot emit all input bytes since the end of the input might be the start of the next boundary
-        val emitEnd = input.length - needle.length - 2
-        if (emitEnd > offset) {
-          emitPartChunk(headers, contentType, input.slice(offset, emitEnd))
-          val simpleEmit: (List[HttpHeader], ContentType, ByteString) ⇒ Unit = (_, _, bytes) ⇒ emit(bytes)
-          continue(input drop emitEnd, 0)(parseEntity(null, null, simpleEmit, simpleEmit))
-        } else continue(input, offset)(parseEntity(headers, contentType, emitPartChunk, emitFinalPartChunk))
-    }
-
-  def emit(bytes: ByteString): Unit = if (bytes.nonEmpty) emit(EntityPart(bytes))
-
-  def emit(element: Output): Unit = output = output.enqueue(element)
 
-  def dequeue(): Output = {
-    val head = output.head
-    output = output.tail
-    head
-  }
-
-  def continue(input: ByteString, offset: Int)(next: (ByteString, Int) ⇒ StateResult): StateResult = {
-    state =
-      math.signum(offset - input.length) match {
-        case -1 ⇒ more ⇒ next(input ++ more, offset)
-        case 0 ⇒ next(_, 0)
-        case 1 ⇒ throw new IllegalStateException
+      def continue(input: ByteString, offset: Int)(next: (ByteString, Int) ⇒ StateResult): StateResult = {
+        state =
+          math.signum(offset - input.length) match {
+            case -1 ⇒ more ⇒ next(input ++ more, offset)
+            case 0 ⇒ next(_, 0)
+            case 1 ⇒ throw new IllegalStateException
+          }
+        done()
       }
-    done()
-  }
 
-  def continue(next: (ByteString, Int) ⇒ StateResult): StateResult = {
-    state = next(_, 0)
-    done()
-  }
+      def continue(next: (ByteString, Int) ⇒ StateResult): StateResult = {
+        state = next(_, 0)
+        done()
+      }
 
-  def fail(summary: String): StateResult = fail(ErrorInfo(summary))
-  def fail(info: ErrorInfo): StateResult = {
-    emit(ParseError(info))
-    terminate()
-  }
+      def fail(summary: String): StateResult = fail(ErrorInfo(summary))
 
-  def terminate(): StateResult = {
-    terminated = true
-    done()
-  }
+      def fail(info: ErrorInfo): StateResult = {
+        emit(ParseError(info))
+        terminate()
+      }
 
-  def done(): StateResult = null // StateResult is a phantom type
+      def terminate(): StateResult = {
+        terminated = true
```

Comment:
I'm not 100 convinced of the name `terminated`. Isn't it more `shouldTerminate`?